src/query.c

```diff
@@ -1605,31 +1605,90 @@ int QueryNode_EvalParams(dict *params, QueryNode *n, QueryError *status) {
   return res;
 }
 
-int QAST_CheckIsValid(QueryAST *q, IndexSpec *spec, RSSearchOptions *opts, QueryError *status) {
-  if (!q || !q->root || !isSpecJson(spec) || !(spec->flags & Index_HasUndefinedOrder)) {
-    return REDISMODULE_OK;
+static int QueryNode_CheckAllowSlopAndInorder(QueryNode *qn, const IndexSpec *spec, bool atTopLevel, QueryError *status) {
+  // Need to check when slop/inorder are locally overridden at query node level, or at query top-level
+  if(atTopLevel || qn->opts.maxSlop >= 0 || (qn->opts.flags & QueryNode_OverriddenInOrder)) {
+    // Check only fields that are used in this query node (either specific fields or all fields)
+    return IndexSpec_CheckAllowSlopAndInorder(spec, qn->opts.fieldMask, status);
+  } else {
+    return 1;
   }
-  return QueryNode_CheckIsValid(q->root, spec, opts, status);
 }
 
-int QueryNode_CheckIsValid(QueryNode *n, IndexSpec *spec, RSSearchOptions *opts, QueryError *status) {
+static inline bool QueryNode_DoesIndexEmpty(QueryNode *n, IndexSpec *spec, RSSearchOptions *opts) {
+  if (opts->flags & QueryNode_IsTag) {
+    return opts->flags & QueryNode_IndexesEmpty;
+  }
+
+  // TEXT field (probably)
+  bool empty_text = n->opts.fieldMask == RS_FIELDMASK_ALL;
+  if (!empty_text) {
+    // Check if there is a field from the field mask that indexes empty. If not,
+    // we throw an error.
+    arrayof(FieldSpec *) fields = IndexSpec_GetFieldsByMask(spec, n->opts.fieldMask);
+    if (array_len(fields) == 0) {
+      // Not a TEXT field. We don't want to throw an error, for backward compatibility.
+      array_free(fields);
+      return true;
+    }
+    array_foreach(fields, fs, {
+      if (FieldSpec_IndexesEmpty(fs)) {
+        empty_text = true;
+        break;
+      }
+    });
+    array_free(fields);
+  }
+  return empty_text;
+}
+
+// If the token is of an empty string, and the searched field doesn't index
+// empty strings, we should return an error
+static inline bool QueryNode_ValidateToken(QueryNode *n, IndexSpec *spec, RSSearchOptions *opts, QueryError *status) {
+  if (n->tn.len == 0 && n->tn.str && !strcmp(n->tn.str, "") && !QueryNode_DoesIndexEmpty(n, spec, opts)) {
+    QueryError_SetError(status, QUERY_ESYNTAX, "Empty value for field that does not index empty values");
```

Comment:
We used to have a more constructive error message such as in commit 10261d2f9
```
QueryError_SetErrorFmt(q->status, QUERY_ENOINDEX,
+                              "Field `%s` should enable `%s` in the index SCHEMA in order to support empty values",
+                              IndexSpec_GetFieldNameByBit(q->sctx->spec, qn->opts.fieldMask),
+                              SPEC_INDEXEMPTY_STR);
```

Reply:
👍 
Changed to a positive message: `QueryError_SetErrorFmt(status, QUERY_ESYNTAX, "Use `%s` in field creation in order to index and query for empty strings", SPEC_INDEXEMPTY_STR);`